torch/nn/modules/module.py

```diff
@@ -1676,16 +1676,16 @@ def remove_from(*dicts_or_sets):
                 modules[name] = value
             else:
                 buffers = self.__dict__.get('_buffers')
-                if buffers is not None and name in buffers:
+                if isinstance(value, Buffer) or buffers is not None and name in buffers:
                     if value is not None and not isinstance(value, torch.Tensor):
                         raise TypeError("cannot assign '{}' as buffer '{}' "
-                                        "(torch.Tensor or None expected)"
+                                        "(torch.nn.Buffer, torch.Tensor or None expected)"
                                         .format(torch.typename(value), name))
-                    for hook in _global_buffer_registration_hooks.values():
-                        output = hook(self, name, value)
-                        if output is not None:
-                            value = output
-                    buffers[name] = value
+                    if isinstance(value, Buffer):
+                        persistent = value.persistent
```

Comment:
First of all, thank you for your very detailed review! Much appreciated! 

Yes you have that correct. I don't think it is entirely necessary since the code in `register_buffer` addresses this here:
https://github.com/pytorch/pytorch/blob/eb4a1a07afecda3a04693c8077c167a3ccf64465/torch/nn/modules/module.py#L538-L541

What I am envisioning is similar to if one called `register_buffer` back to back with `persistent` set to different values in the 2 calls. This currently works in the current implementation without throwing an assert so I'm just trying to replicate that behavior.

Reply:
Hm ok this seems fair